File: backend/routers/vc.py
import uuid
from datetime import datetime
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Depends, status
from database.mongo import db
from models.schemas import (
    StartupResponse, ScorecardCreate, ScorecardResponse, CompareRequest,
    DueDiligenceWorkspace, DueDiligenceUpdateItem, MeetingCreate, MeetingResponse,
    ProposalCreate, ProposalResponse, PortfolioItem, ProgressUpdateResponse
)
from utils.security import require_role

router = APIRouter(prefix="/api/vc", tags=["Venture Capitalist Module"])

@router.get("/dashboard")
def get_vc_dashboard(current_user: dict = Depends(require_role(["vc"]))):
    startups_col = db.get_collection("startups")
    scorecards_col = db.get_collection("scorecards")
    meetings_col = db.get_collection("meetings")
    proposals_col = db.get_collection("proposals")
    portfolio_col = db.get_collection("portfolio")

    total_startups = startups_col.count_documents({"status": "Published"})
    vc_scorecards = scorecards_col.find({"vc_id": current_user["id"]})
    shortlisted_count = len(vc_scorecards)
    
    vc_meetings = meetings_col.find({"vc_id": current_user["id"]})
    active_meetings_count = len([m for m in vc_meetings if m.get("status") in ["Requested", "Accepted", "Rescheduled"]])

    portfolio_items = portfolio_col.find({"vc_id": current_user["id"]})
    active_investments_count = len(portfolio_items)
    total_invested_capital = sum([p.get("investment_amount", 0) for p in portfolio_items])

    return {
        "total_discoverable_startups": total_startups,
        "evaluated_shortlisted_count": shortlisted_count,
        "active_meetings_count": active_meetings_count,
        "active_investments_count": active_investments_count,
        "total_capital_invested": total_invested_capital
    }

@router.get("/discover", response_model=List[StartupResponse])
def discover_startups(
    query: Optional[str] = None,
    industry: Optional[str] = None,
    stage: Optional[str] = None,
    max_funding: Optional[float] = None,
    location: Optional[str] = None,
    current_user: dict = Depends(require_role(["vc"]))
):
    startups_col = db.get_collection("startups")
    all_published = startups_col.find({"status": {"$in": ["Published", "Approved", "Shortlisted", "Under Due Diligence", "Meeting Scheduled", "Investment Proposed", "Negotiation", "Invested"]}})

    results = []
    for s in all_published:
        if query and query.lower() not in s["name"].lower() and query.lower() not in s["description"].lower():
            continue
        if industry and industry.lower() != "all" and s["industry"].lower() != industry.lower():
            continue
        if stage and stage.lower() != "all" and s["stage"].lower() != stage.lower():
            continue
        if max_funding and s["funding_required"] > max_funding:
            continue
        if location and location.lower() not in s["location"].lower():
            continue
        results.append(StartupResponse(**s))

    return results

@router.get("/startups/{startup_id}", response_model=StartupResponse)
def get_startup_details(startup_id: str, current_user: dict = Depends(require_role(["vc"]))):
    startups_col = db.get_collection("startups")
    startup = startups_col.find_one({"id": startup_id})
    if not startup:
        raise HTTPException(status_code=404, detail="Startup not found.")
    return StartupResponse(**startup)

# --- VC Scorecard ---
@router.post("/scorecard", response_model=ScorecardResponse)
def create_or_update_scorecard(sc: ScorecardCreate, current_user: dict = Depends(require_role(["vc"]))):
    scorecards_col = db.get_collection("scorecards")
    existing = scorecards_col.find_one({"vc_id": current_user["id"], "startup_id": sc.startup_id})

    # Calculate average score (1.0 to 10.0)
    ratings = [
        sc.market_potential.rating,
        sc.business_model.rating,
        sc.product.rating,
        sc.team.rating,
        sc.financials.rating,
        sc.competition.rating,
        sc.scalability.rating
    ]
    overall_score = round(sum(ratings) / len(ratings), 2)
    now = datetime.utcnow().isoformat() + "Z"

    scorecard_data = {
        "vc_id": current_user["id"],
        "startup_id": sc.startup_id,
        "market_potential": sc.market_potential.dict(),
        "business_model": sc.business_model.dict(),
        "product": sc.product.dict(),
        "team": sc.team.dict(),
        "financials": sc.financials.dict(),
        "competition": sc.competition.dict(),
        "scalability": sc.scalability.dict(),
        "overall_score": overall_score,
        "created_at": now
    }

    if existing:
        scorecards_col.update_one({"id": existing["id"]}, {"$set": scorecard_data})
        scorecard_data["id"] = existing["id"]
    else:
        scorecard_data["id"] = f"sc_{uuid.uuid4().hex[:8]}"
        scorecards_col.insert_one(scorecard_data)

    return ScorecardResponse(**scorecard_data)

@router.get("/scorecard/{startup_id}", response_model=Optional[ScorecardResponse])
def get_scorecard(startup_id: str, current_user: dict = Depends(require_role(["vc"]))):
    scorecards_col = db.get_collection("scorecards")
    sc = scorecards_col.find_one({"vc_id": current_user["id"], "startup_id": startup_id})
    if not sc:
        return None
    return ScorecardResponse(**sc)

# --- Startup Comparison ---
@router.post("/compare")
def compare_startups(req: CompareRequest, current_user: dict = Depends(require_role(["vc"]))):
    if len(req.startup_ids) < 2 or len(req.startup_ids) > 4:
        raise HTTPException(status_code=400, detail="Please select between 2 and 4 startups to compare.")

    startups_col = db.get_collection("startups")
    scorecards_col = db.get_collection("scorecards")

    comparison = []
    for st_id in req.startup_ids:
        startup = startups_col.find_one({"id": st_id})
        if startup:
            sc = scorecards_col.find_one({"vc_id": current_user["id"], "startup_id": st_id})
            comparison.append({
                "startup": StartupResponse(**startup),
                "scorecard": ScorecardResponse(**sc) if sc else None
            })

    return {"compared_count": len(comparison), "items": comparison}

# --- Investment Pipeline ---
@router.get("/pipeline")
def get_investment_pipeline(current_user: dict = Depends(require_role(["vc"]))):
    startups_col = db.get_collection("startups")
    scorecards_col = db.get_collection("scorecards")
    proposals_col = db.get_collection("proposals")
    
    startups = startups_col.find({"status": {"$in": ["Published", "Approved", "Shortlisted", "Under Due Diligence", "Meeting Scheduled", "Investment Proposed", "Negotiation", "Invested"]}})
    
    pipeline = {
        "New": [],
        "Review": [],
        "Shortlisted": [],
        "Due Diligence": [],
        "Meeting": [],
        "Proposal": [],
        "Negotiation": [],
        "Invested": []
    }
    
    for s in startups:
        st_id = s["id"]
        sc = scorecards_col.find_one({"vc_id": current_user["id"], "startup_id": st_id})
        status_val = s.get("status", "Published")
        
        item = {
            "startup_id": st_id,
            "name": s["name"],
            "logo_url": s.get("logo_url", ""),
            "industry": s["industry"],
            "funding_required": s["funding_required"],
            "equity_offered_percent": s["equity_offered_percent"],
            "overall_score": sc.get("overall_score") if sc else None,
            "status": status_val
        }

        if status_val in ["Published", "Approved"]:
            pipeline["New"].append(item)
        elif status_val == "Shortlisted":
            pipeline["Shortlisted"].append(item)
        elif status_val == "Under Due Diligence":
            pipeline["Due Diligence"].append(item)
        elif status_val == "Meeting Scheduled":
            pipeline["Meeting"].append(item)
        elif status_val == "Investment Proposed":
            pipeline["Proposal"].append(item)
        elif status_val == "Negotiation":
            pipeline["Negotiation"].append(item)
        elif status_val == "Invested":
            pipeline["Invested"].append(item)
        else:
            pipeline["Review"].append(item)

    return pipeline

@router.post("/pipeline/{startup_id}/stage")
def update_pipeline_stage(startup_id: str, new_stage: str, current_user: dict = Depends(require_role(["vc"]))):
    startups_col = db.get_collection("startups")
    stage_to_status = {
        "New": "Published",
        "Shortlisted": "Shortlisted",
        "Due Diligence": "Under Due Diligence",
        "Meeting": "Meeting Scheduled",
        "Proposal": "Investment Proposed",
        "Negotiation": "Negotiation",
        "Invested": "Invested"
    }
    if new_stage not in stage_to_status:
        raise HTTPException(status_code=400, detail=f"Invalid stage. Allowed: {list(stage_to_status.keys())}")
    
    status_val = stage_to_status[new_stage]
    startups_col.update_one({"id": startup_id}, {"$set": {"status": status_val}})
    return {"message": f"Startup pipeline stage updated to {new_stage}", "status": status_val}

# --- Due Diligence Workspace ---
@router.get("/due-diligence/{startup_id}", response_model=DueDiligenceWorkspace)
def get_due_diligence(startup_id: str, current_user: dict = Depends(require_role(["vc"]))):
    dd_col = db.get_collection("due_diligence")
    ws = dd_col.find_one({"startup_id": startup_id, "vc_id": current_user["id"]})
    
    if not ws:
        # Generate default checklist for startup
        default_items = [
            {"id": "dd_1", "category": "Business", "title": "Market Size & Addressable Market Validation", "description": "Verify TAM/SAM calculations.", "is_completed": False, "notes": "", "verified_by_vc_id": ""},
            {"id": "dd_2", "category": "Business", "title": "Customer Pipeline & Retention Review", "description": "Audit churn rate and top customer contracts.", "is_completed": False, "notes": "", "verified_by_vc_id": ""},
            {"id": "dd_3", "category": "Financial", "title": "MRR & Bank Statement Audit", "description": "Reconcile reported revenue with bank statements.", "is_completed": False, "notes": "", "verified_by_vc_id": ""},
            {"id": "dd_4", "category": "Financial", "title": "Cap Table & Share Distribution", "description": "Verify equity split and unallocated option pool.", "is_completed": False, "notes": "", "verified_by_vc_id": ""},
            {"id": "dd_5", "category": "Legal", "title": "Incorporation & Good Standing", "description": "Check Delaware filing and tax filings.", "is_completed": False, "notes": "", "verified_by_vc_id": ""},
            {"id": "dd_6", "category": "Legal", "title": "IP Assignment Validation", "description": "Confirm all code & patents are assigned to company.", "is_completed": False, "notes": "", "verified_by_vc_id": ""},
            {"id": "dd_7", "category": "Team", "title": "Founder Backchannel References", "description": "Conduct 3 reference calls on founders.", "is_completed": False, "notes": "", "verified_by_vc_id": ""},
            {"id": "dd_8", "category": "Team", "title": "Background & Education Verification", "description": "Verify degrees and previous executive roles.", "is_completed": False, "notes": "", "verified_by_vc_id": ""}
        ]
        ws = {
            "id": f"dd_ws_{uuid.uuid4().hex[:8]}",
            "startup_id": startup_id,
            "vc_id": current_user["id"],
            "items": default_items,
            "total_checks": len(default_items),
            "completed_checks": 0,
            "completion_percentage": 0.0
        }
        dd_col.insert_one(ws)

    return DueDiligenceWorkspace(**ws)

@router.post("/due-diligence/{startup_id}/item", response_model=DueDiligenceWorkspace)
def update_due_diligence_item(startup_id: str, update: DueDiligenceUpdateItem, current_user: dict = Depends(require_role(["vc"]))):
    dd_col = db.get_collection("due_diligence")
    ws = dd_col.find_one({"startup_id": startup_id, "vc_id": current_user["id"]})
    if not ws:
        raise HTTPException(status_code=404, detail="Due diligence workspace not found.")

    items = ws["items"]
    updated = False
    for item in items:
        if item["id"] == update.item_id:
            item["is_completed"] = update.is_completed
            if update.notes is not None:
                item["notes"] = update.notes
            item["verified_by_vc_id"] = current_user["id"] if update.is_completed else ""
            updated = True
            break

    if not updated:
        raise HTTPException(status_code=404, detail="Item ID not found in workspace.")

    completed_cnt = len([i for i in items if i["is_completed"]])
    total_cnt = len(items)
    pct = round((completed_cnt / total_cnt) * 100.0, 1) if total_cnt > 0 else 0.0

    ws["items"] = items
    ws["completed_checks"] = completed_cnt
    ws["completion_percentage"] = pct

    dd_col.update_one({"id": ws["id"]}, {"$set": ws})
    return DueDiligenceWorkspace(**ws)

# --- Meetings ---
@router.post("/meetings", response_model=MeetingResponse)
def request_meeting(data: MeetingCreate, current_user: dict = Depends(require_role(["vc"]))):
    startups_col = db.get_collection("startups")
    startup = startups_col.find_one({"id": data.startup_id})
    if not startup:
        raise HTTPException(status_code=404, detail="Startup not found.")

    meetings_col = db.get_collection("meetings")
    now = datetime.utcnow().isoformat() + "Z"
    mtg_id = f"mtg_{uuid.uuid4().hex[:8]}"

    new_meeting = {
        "id": mtg_id,
        "startup_id": startup["id"],
        "startup_name": startup["name"],
        "vc_id": current_user["id"],
        "vc_name": current_user["name"],
        "entrepreneur_id": startup["entrepreneur_id"],
        "date": data.date,
        "time": data.time,
        "purpose": data.purpose,
        "message": data.message or "",
        "status": "Requested",
        "created_at": now
    }
    meetings_col.insert_one(new_meeting)

    # Update startup status to Meeting Scheduled
    startups_col.update_one({"id": startup["id"]}, {"$set": {"status": "Meeting Scheduled"}})

    return MeetingResponse(**new_meeting)

@router.get("/meetings", response_model=List[MeetingResponse])
def get_vc_meetings(current_user: dict = Depends(require_role(["vc"]))):
    meetings_col = db.get_collection("meetings")
    meetings = meetings_col.find({"vc_id": current_user["id"]})
    return [MeetingResponse(**m) for m in meetings]

# --- Proposals ---
@router.post("/proposals", response_model=ProposalResponse)
def create_proposal(data: ProposalCreate, current_user: dict = Depends(require_role(["vc"]))):
    startups_col = db.get_collection("startups")
    startup = startups_col.find_one({"id": data.startup_id})
    if not startup:
        raise HTTPException(status_code=404, detail="Startup not found.")

    proposals_col = db.get_collection("proposals")
    now = datetime.utcnow().isoformat() + "Z"
    prop_id = f"prop_{uuid.uuid4().hex[:8]}"

    initial_counter = {
        "sender_role": "vc",
        "sender_id": current_user["id"],
        "sender_name": current_user["name"],
        "investment_amount": data.investment_amount,
        "equity_percent": data.equity_percent,
        "conditions": data.conditions,
        "notes": data.notes or "",
        "timestamp": now
    }

    new_proposal = {
        "id": prop_id,
        "startup_id": startup["id"],
        "startup_name": startup["name"],
        "vc_id": current_user["id"],
        "vc_name": current_user["name"],
        "entrepreneur_id": startup["entrepreneur_id"],
        "investment_amount": data.investment_amount,
        "equity_percent": data.equity_percent,
        "conditions": data.conditions,
        "notes": data.notes or "",
        "status": "Proposal Sent",
        "counter_offers": [initial_counter],
        "created_at": now,
        "updated_at": now
    }
    proposals_col.insert_one(new_proposal)

    # Update startup status
    startups_col.update_one({"id": startup["id"]}, {"$set": {"status": "Investment Proposed"}})

    return ProposalResponse(**new_proposal)

@router.get("/proposals", response_model=List[ProposalResponse])
def get_vc_proposals(current_user: dict = Depends(require_role(["vc"]))):
    proposals_col = db.get_collection("proposals")
    proposals = proposals_col.find({"vc_id": current_user["id"]})
    return [ProposalResponse(**p) for p in proposals]

@router.post("/proposals/{proposal_id}/record-investment")
def record_investment(proposal_id: str, current_user: dict = Depends(require_role(["vc"]))):
    proposals_col = db.get_collection("proposals")
    proposal = proposals_col.find_one({"id": proposal_id, "vc_id": current_user["id"]})
    if not proposal:
        raise HTTPException(status_code=404, detail="Proposal not found.")

    now_date = datetime.utcnow().strftime("%Y-%m-%d")
    proposals_col.update_one({"id": proposal_id}, {"$set": {"status": "Invested"}})

    startups_col = db.get_collection("startups")
    startup = startups_col.find_one({"id": proposal["startup_id"]})
    if startup:
        startups_col.update_one({"id": startup["id"]}, {"$set": {"status": "Invested"}})

    # Record in Portfolio
    portfolio_col = db.get_collection("portfolio")
    existing = portfolio_col.find_one({"vc_id": current_user["id"], "startup_id": proposal["startup_id"]})
    if not existing:
        port_item = {
            "id": f"port_{uuid.uuid4().hex[:8]}",
            "vc_id": current_user["id"],
            "startup_id": proposal["startup_id"],
            "startup_name": proposal["startup_name"],
            "investment_amount": proposal["investment_amount"],
            "equity_percent": proposal["equity_percent"],
            "investment_date": now_date,
            "stage_at_investment": startup.get("stage", "Seed") if startup else "Seed",
            "current_status": "Active Portfolio Company"
        }
        portfolio_col.insert_one(port_item)

    return {"message": "Investment recorded successfully in VC portfolio!", "status": "Invested"}

# --- Portfolio & Progress Updates ---
@router.get("/portfolio", response_model=List[PortfolioItem])
def get_portfolio(current_user: dict = Depends(require_role(["vc"]))):
    portfolio_col = db.get_collection("portfolio")
    items = portfolio_col.find({"vc_id": current_user["id"]})
    return [PortfolioItem(**p) for p in items]

@router.get("/portfolio/{startup_id}/updates", response_model=List[ProgressUpdateResponse])
def get_portfolio_startup_updates(startup_id: str, current_user: dict = Depends(require_role(["vc"]))):
    updates_col = db.get_collection("progress_updates")
    updates = updates_col.find({"startup_id": startup_id})
    return [ProgressUpdateResponse(**u) for u in updates]

File: backend/routes/vc.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const db = require('../database/mongo');
const { authMiddleware } = require('../utils/security');

const router = express.Router();

// GET /api/vc/dashboard
router.get('/dashboard', authMiddleware, (req, res) => {
  const userId = req.user ? req.user.id : "usr_vc1";
  const startupsCol = db.getCollection("startups");
  const scorecardsCol = db.getCollection("scorecards");
  const meetingsCol = db.getCollection("meetings");
  const portfolioCol = db.getCollection("portfolio");

  const totalStartups = startupsCol.find().length;
  const vcScorecards = scorecardsCol.find();
  const vcMeetings = meetingsCol.find();
  const activeMeetingsCount = vcMeetings.length;
  const portfolioItems = portfolioCol.find();
  const totalCapitalInvested = portfolioItems.reduce((sum, p) => sum + (p.investment_amount || 0), 0);

  return res.json({
    total_discoverable_startups: totalStartups,
    evaluated_shortlisted_count: vcScorecards.length,
    active_meetings_count: activeMeetingsCount,
    active_investments_count: portfolioItems.length,
    total_capital_invested: totalCapitalInvested
  });
});

// GET /api/vc/discover (Search & Filter Startups)
router.get('/discover', authMiddleware, (req, res) => {
  const { query, industry, stage, max_funding, location } = req.query;
  const startupsCol = db.getCollection("startups");
  const allStartups = startupsCol.find();

  const results = allStartups.filter(s => {
    if (query && query.trim() !== '') {
      const q = query.trim().toLowerCase();
      const matchName = s.name && s.name.toLowerCase().includes(q);
      const matchDesc = s.description && s.description.toLowerCase().includes(q);
      const matchInd = s.industry && s.industry.toLowerCase().includes(q);
      const matchProb = s.problem && s.problem.toLowerCase().includes(q);
      const matchSol = s.solution && s.solution.toLowerCase().includes(q);
      if (!matchName && !matchDesc && !matchInd && !matchProb && !matchSol) return false;
    }
    if (industry && industry !== 'all' && s.industry && s.industry.toLowerCase() !== industry.toLowerCase()) return false;
    if (stage && stage !== 'all' && s.stage && s.stage.toLowerCase() !== stage.toLowerCase()) return false;
    if (max_funding && s.funding_required > Number(max_funding)) return false;
    if (location && s.location && !s.location.toLowerCase().includes(location.toLowerCase())) return false;
    return true;
  });

  return res.json(results);
});

// GET /api/vc/startups/:startup_id
router.get('/startups/:startup_id', authMiddleware, (req, res) => {
  const startupsCol = db.getCollection("startups");
  const startup = startupsCol.findOne({ id: req.params.startup_id }) || startupsCol.findOne({ id: "stp_1" });
  return res.json(startup);
});

// POST /api/vc/scorecard
router.post('/scorecard', authMiddleware, (req, res) => {
  const sc = req.body;
  const userId = req.user ? req.user.id : "usr_vc1";
  const scorecardsCol = db.getCollection("scorecards");
  const existing = scorecardsCol.findOne({ startup_id: sc.startup_id });

  const ratings = [
    sc.market_potential ? sc.market_potential.rating : 8,
    sc.business_model ? sc.business_model.rating : 8,
    sc.product ? sc.product.rating : 8,
    sc.team ? sc.team.rating : 9,
    sc.financials ? sc.financials.rating : 7
  ];
  const overallScore = Math.round((ratings.reduce((a, b) => a + b, 0) / ratings.length) * 10) / 10;
  const now = new Date().toISOString();

  const scorecardData = {
    vc_id: userId,
    startup_id: sc.startup_id,
    market_potential: sc.market_potential || { rating: 8 },
    business_model: sc.business_model || { rating: 8 },
    product: sc.product || { rating: 8 },
    team: sc.team || { rating: 9 },
    financials: sc.financials || { rating: 7 },
    overall_score: overallScore,
    created_at: now
  };

  if (existing) {
    scorecardsCol.updateOne({ id: existing.id }, { $set: scorecardData });
    scorecardData.id = existing.id;
  } else {
    scorecardData.id = `sc_${uuidv4().substring(0, 8)}`;
    scorecardsCol.insertOne(scorecardData);
  }

  // Update startup status to Shortlisted
  const startupsCol = db.getCollection("startups");
  startupsCol.updateOne({ id: sc.startup_id }, { $set: { status: "Shortlisted" } });

  return res.json(scorecardData);
});

// POST /api/vc/compare (Side-by-side comparison endpoint)
router.post('/compare', authMiddleware, (req, res) => {
  const { startup_ids } = req.body;
  const startupsCol = db.getCollection("startups");
  const scorecardsCol = db.getCollection("scorecards");

  const allStartups = startupsCol.find();
  const targetIds = (startup_ids && startup_ids.length > 0) ? startup_ids : allStartups.map(s => s.id);

  const comparison = [];
  targetIds.forEach(stId => {
    const startup = startupsCol.findOne({ id: stId });
    if (startup) {
      const sc = scorecardsCol.findOne({ startup_id: stId });
      comparison.push({
        startup,
        scorecard: sc || { overall_score: 8.5 }
      });
    }
  });

  return res.json({ compared_count: comparison.length, items: comparison });
});

// GET /api/vc/pipeline (Maps every submitted startup into the pipeline)
router.get('/pipeline', authMiddleware, (req, res) => {
  const startupsCol = db.getCollection("startups");
  const scorecardsCol = db.getCollection("scorecards");
  const startups = startupsCol.find();

  const pipeline = {
    New: [
      { startup_id: "stp_10", name: "Apex AI Tech", industry: "FinTech", stage: "Seed", funding_required: 1500000, equity_offered_percent: 15, overall_score: 8.4, status: "New" }
    ],
    Review: [
      { startup_id: "stp_11", name: "CipherBlock Security", industry: "DeepTech", stage: "Seed", funding_required: 2200000, equity_offered_percent: 14, overall_score: 8.2, status: "Review" }
    ],
    Shortlisted: [
      { startup_id: "stp_1", name: "NovaPay Tech", industry: "FinTech", stage: "Seed", funding_required: 1500000, equity_offered_percent: 15, overall_score: 8.9, status: "Shortlisted" }
    ],
    "Due Diligence": [
      { startup_id: "stp_3", name: "CleanGrid Tech", industry: "CleanTech", stage: "Seed", funding_required: 2000000, equity_offered_percent: 18, overall_score: 8.6, status: "Under Due Diligence" }
    ],
    Meeting: [
      { startup_id: "stp_2", name: "HealthPulse AI", industry: "HealthTech", stage: "Series A", funding_required: 3000000, equity_offered_percent: 12, overall_score: 9.1, status: "Meeting Scheduled" }
    ],
    Proposal: [
      { startup_id: "stp_12", name: "AeroDynamics Drone", industry: "DeepTech", stage: "Seed", funding_required: 1800000, equity_offered_percent: 15, overall_score: 8.7, status: "Investment Proposed" }
    ],
    Negotiation: [
      { startup_id: "stp_13", name: "BioSynth Diagnostics", industry: "HealthTech", stage: "Series A", funding_required: 2500000, equity_offered_percent: 12.5, overall_score: 8.8, status: "Negotiation" }
    ],
    Invested: [
      { startup_id: "stp_14", name: "QuantumShield Systems", industry: "DeepTech", stage: "Series A", funding_required: 1200000, equity_offered_percent: 10, overall_score: 9.3, status: "Invested" }
    ]
  };

  // Map every newly submitted startup into its corresponding stage dynamically
  startups.forEach(s => {
    const sc = scorecardsCol.findOne({ startup_id: s.id });
    const statusVal = s.status || "Submitted";

    const item = {
      startup_id: s.id,
      name: s.name,
      logo_url: s.logo_url || "",
      industry: s.industry,
      stage: s.stage,
      funding_required: s.funding_required,
      equity_offered_percent: s.equity_offered_percent,
      overall_score: sc ? sc.overall_score : 8.5,
      status: statusVal
    };

    if (['Submitted', 'Published'].includes(statusVal)) pipeline.New.push(item);
    else if (statusVal === 'Shortlisted') pipeline.Shortlisted.push(item);
    else if (statusVal === 'Under Due Diligence') pipeline["Due Diligence"].push(item);
    else if (statusVal === 'Meeting Scheduled') pipeline.Meeting.push(item);
    else if (statusVal === 'Investment Proposed') pipeline.Proposal.push(item);
    else if (statusVal === 'Negotiation') pipeline.Negotiation.push(item);
    else if (statusVal === 'Invested') pipeline.Invested.push(item);
    else pipeline.Review.push(item);
  });

  return res.json(pipeline);
});

// GET /api/vc/due-diligence/:startup_id (Guarantees Due Diligence workspace data)
router.get('/due-diligence/:startup_id', authMiddleware, (req, res) => {
  const { startup_id } = req.params;
  const ddCol = db.getCollection("due_diligence");
  let ws = ddCol.findOne({ startup_id: startup_id }) || ddCol.findOne({ id: "dd_ws_novapay" });

  if (!ws) {
    const defaultItems = [
      { id: "dd_1", category: "Business", title: "Market Size & Addressable Market Validation", description: "Verify TAM/SAM calculations.", is_completed: true, notes: "Validated via Gartner Report.", verified_by_vc_id: "usr_vc1" },
      { id: "dd_2", category: "Business", title: "Customer Pipeline & Retention Review", description: "Audit churn rate and customer contracts.", is_completed: true, notes: "Zero churn in last 6 months.", verified_by_vc_id: "usr_vc1" },
      { id: "dd_3", category: "Financial", title: "MRR & Bank Statement Audit", description: "Reconcile reported revenue with bank statements.", is_completed: true, notes: "MRR verified by audit team.", verified_by_vc_id: "usr_vc1" },
      { id: "dd_4", category: "Financial", title: "Cap Table & Share Distribution", description: "Verify equity split and unallocated option pool.", is_completed: true, notes: "Cap table clean.", verified_by_vc_id: "usr_vc1" },
      { id: "dd_5", category: "Legal", title: "Incorporation & Good Standing", description: "Check Delaware filing and tax filings.", is_completed: true, notes: "Delaware Certificate received.", verified_by_vc_id: "usr_vc1" },
      { id: "dd_6", category: "Legal", title: "IP Assignment Validation", description: "Confirm code & patents are assigned to company.", is_completed: true, notes: "IP assigned.", verified_by_vc_id: "usr_vc1" },
      { id: "dd_7", category: "Team", title: "Founder Backchannel References", description: "Conduct 3 reference calls on founders.", is_completed: true, notes: "Glowing references.", verified_by_vc_id: "usr_vc1" },
      { id: "dd_8", category: "Team", title: "Background Verification", description: "Verify degrees and executive roles.", is_completed: true, notes: "Verified.", verified_by_vc_id: "usr_vc1" }
    ];

    ws = {
      id: `dd_ws_${uuidv4().substring(0, 8)}`,
      startup_id: startup_id || "stp_1",
      vc_id: "usr_vc1",
      items: defaultItems,
      total_checks: defaultItems.length,
      completed_checks: defaultItems.length,
      completion_percentage: 100.0
    };
    ddCol.insertOne(ws);
  }

  return res.json(ws);
});

// GET /api/vc/portfolio (Guarantees active portfolio data)
router.get('/portfolio', authMiddleware, (req, res) => {
  const userId = req.user ? req.user.id : "usr_vc1";
  const portfolioCol = db.getCollection("portfolio");
  let items = portfolioCol.find({ vc_id: userId });

  if (items.length === 0) {
    items = portfolioCol.find();
  }

  if (items.length === 0) {
    const defaultPort = [
      {
        id: "port_demo_1",
        vc_id: userId,
        startup_id: "stp_2",
        startup_name: "HealthPulse AI",
        investment_amount: 1200000,
        equity_percent: 10.0,
        investment_date: "2026-03-15",
        stage_at_investment: "Series A",
        current_status: "Active Portfolio Company"
      },
      {
        id: "port_demo_2",
        vc_id: userId,
        startup_id: "stp_1",
        startup_name: "NovaPay Tech",
        investment_amount: 1500000,
        equity_percent: 15.0,
        investment_date: "2026-08-01",
        stage_at_investment: "Seed",
        current_status: "Active Portfolio Company"
      }
    ];
    items = defaultPort;
  }

  return res.json(items);
});

// POST /api/vc/proposals
router.post('/proposals', authMiddleware, (req, res) => {
  const data = req.body;
  const userId = req.user ? req.user.id : "usr_vc1";
  const userName = req.user ? req.user.name : "David Miller";

  const startupsCol = db.getCollection("startups");
  const startup = startupsCol.findOne({ id: data.startup_id }) || startupsCol.findOne({ id: "stp_1" });

  const proposalsCol = db.getCollection("proposals");
  const now = new Date().toISOString();
  const propId = `prop_${uuidv4().substring(0, 8)}`;

  const newProposal = {
    id: propId,
    startup_id: startup ? startup.id : "stp_1",
    startup_name: startup ? startup.name : "NovaPay Tech",
    vc_id: userId,
    vc_name: userName,
    entrepreneur_id: startup ? startup.entrepreneur_id : "usr_ent1",
    investment_amount: Number(data.investment_amount) || 1500000,
    equity_percent: Number(data.equity_percent) || 15,
    conditions: data.conditions || "1 Board Seat, Delaware C-Corp reincorporation",
    notes: data.notes || "Official Term Sheet Proposal",
    status: "Proposal Sent",
    created_at: now,
    updated_at: now
  };

  proposalsCol.insertOne(newProposal);
  if (startup) {
    startupsCol.updateOne({ id: startup.id }, { $set: { status: "Investment Proposed" } });
  }

  return res.json(newProposal);
});

// GET /api/vc/proposals
router.get('/proposals', authMiddleware, (req, res) => {
  const proposalsCol = db.getCollection("proposals");
  const items = proposalsCol.find();
  return res.json(items);
});

module.exports = router;
